pkg/controller/direct/gkebackup/backupplan_controller.go

```diff
@@ -0,0 +1,330 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +tool:controller
+// proto.service: google.cloud.gkebackup.v1.BackupForGKE
+// proto.message: google.cloud.gkebackup.v1.BackupPlan
+// crd.type: GKEBackupBackupPlan
+// crd.version: v1alpha1
+
+package gkebackup
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	gcp "cloud.google.com/go/gkebackup/apiv1"
+	"cloud.google.com/go/gkebackup/apiv1/gkebackuppb"
+	pb "cloud.google.com/go/gkebackup/apiv1/gkebackuppb"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/fieldmaskpb"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	krm "github.com/GoogleCloudPlatform/k8s-config-connector/apis/gkebackup/v1alpha1"
+	refs "github.com/GoogleCloudPlatform/k8s-config-connector/apis/refs/v1beta1"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/config"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct/directbase"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct/registry"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/resourceoverrides"
+)
+
+func init() {
+	registry.RegisterModel(krm.GKEBackupBackupPlanGVK, NewBackupPlanModel)
+}
+
+func NewBackupPlanModel(ctx context.Context, config *config.ControllerConfig) (directbase.Model, error) {
+	return &backupPlanModel{config: *config}, nil
+}
+
+var _ directbase.Model = &backupPlanModel{}
+
+type backupPlanModel struct {
+	config config.ControllerConfig
+}
+
+func (m *backupPlanModel) AdapterForObject(ctx context.Context, reader client.Reader, u *unstructured.Unstructured) (directbase.Adapter, error) {
+	obj := &krm.GKEBackupBackupPlan{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &obj); err != nil {
+		return nil, fmt.Errorf("error converting to %T: %w", obj, err)
+	}
+
+	id, err := krm.NewBackupPlanIdentity(ctx, reader, obj)
+	if err != nil {
+		return nil, err
+	}
+
+	// normalize reference fields
+	if obj.Spec.ClusterRef != nil {
+		if _, err := obj.Spec.ClusterRef.NormalizedExternal(ctx, reader, obj.GetNamespace()); err != nil {
+			return nil, err
+		}
+	}
+	if obj.Spec.BackupConfig != nil &&
+		obj.Spec.BackupConfig.EncryptionKey != nil &&
+		obj.Spec.BackupConfig.EncryptionKey.KMSKeyRef != nil {
+		if _, err := refs.ResolveKMSCryptoKeyRef(ctx, reader, obj, obj.Spec.BackupConfig.EncryptionKey.KMSKeyRef); err != nil {
+			return nil, err
+		}
+	}
```

Comment:
I think we don't want to put these normalization here if they are optional (it blocks deletion)

Reply:
I will review the 10 controllers I added over the past week related to this issue.